### ISSUE TITLE
[Loot] `extra.items` JSON includes irrelevant loot, ignoring denylist

### ISSUE THREAD

**Issue by tmgrhm** [pajlads/DinkPlugin#629]
### Checklist

- [X] I'm reporting a problem with Dink
- [X] I've searched the [issues and pull requests](https://github.com/pajlads/Dinkplugin/issues?q=) for similar looking reports

### Describe your issue

Unsure if this is truly a bug — I assume at this point it's not but I'd like to at least understand why and potentially find a solution — but when receiving the payload JSON from a `LOOT` trigger, the `extra.items` array includes items irrelevant to the trigger if the drop included multiple items. This is true even if an item is on the denylist. 
These items are not included in the `embed` discord message, as expected,  so why are they not also filtered out of the `extra.items` content?

For example, I have an Allowlist of `Raw chicken` and kill a Chicken. The Discord message correctly shows that I got `1 x Raw Chicken`, while the JSON tells me I got [`1 x Bones`, `1 x Raw Chicken`]. I don't care about the bones, and this actually makes it a headache for me to process this data as I now need to keep a local Allowlist of items and check against it before processing the item data further. The Bones still show up in `extra.items` even if my Denylist includes `Bones`.

For context, I'm using Dink for a clan bingo event to create a list of Loot drops (in a Google Sheet, which is then processed via Google Apps Script to record, calculate and display tile progress, points, etc). There'll be hundreds of items needed from sources all across the game so even if the Denylist was respected that would not be a practical solution for me as I'd need to include every single other possible drop from the monsters who drop the items which _are_ wanted. The other solution would be to manually parse the `embed`, as this is already filtered, but that would require mandating a standard format for the message and seems sort of counterproductive when Dink already sends the JSON.

### Screenshots

<img width="242" alt="image" src="https://github.com/user-attachments/assets/412f3b73-a230-4002-b417-0b32d1849577" />

<img width="490" alt="image" src="https://github.com/user-attachments/assets/15ded33f-f059-4c49-ac6d-58272f00221b" />

**See `extra.items[0]`:**

<img width="1109" alt="image" src="https://github.com/user-attachments/assets/5db06f36-d5e0-42b7-b963-b1e50da237b1" />

### Runelite Logs

N/A

### Runelite Version

1.10.46

**Comment by iProdigy** [comment]
`extra.items` being comprehensive is intentional for third-party handlers to use as desired, but we could add a json field that indicates whether the item was on the allow/deny list

**Comment by iProdigy** [comment]
> I'd need to include every single other possible drop from the monsters who drop the items which are wanted

I don't follow, can't you use `source` if you want to see if a given mob was killed, regardless of what specifically dropped? (& for the specific items you care about, just compare with your allowlist)

**Comment by tmgrhm** [comment]
>but we could add a json field that indicates whether the item was on the allow/deny list

That would be fantastic, if it's not too much trouble!  
I expect to be using the allowlist for everything so that works for me but for what it's worth I don't think that solves the same issue where you're relying on the value/rarity to filter, unless I'm missing something?  
Can't think of a good name but a flag along the lines of `filtered`/`fitsCriteria` which is `true` for all items included in the `embed` description and otherwise `false`… or a `reason` enum for `allowlist`/`value`/`rarity`/`none`/`denylist`? Idk, might be overcomplicating it

>I don't follow, can't you use source if you want to see if a given mob was killed, regardless of what specifically dropped? (& for the specific items you care about, just compare with your allowlist)

Basically that was a hypothetical where the denylist is respected by the `extras.items` list, meaning, in the `Loot` config, you could exclude every item from an npc except those items which you're looking for (I'm looking for items dropped, not which mob was killed, though that will be relevant for some tiles (e.g. getting a Dragon Pickaxe from KQ and checking it wasn't from a wildy boss) — but it doesn't matter, the flag for items being on each list solves this 🙂

**Comment by iProdigy** [comment]
Created PR that adds a `criteria` field (EnumSet, which is serialized as an array of strings) for each item that contains `VALUE` is item exceeds value threshold, `RARITY` if item meets rarity threshold, `ALLOWLIST` if item is on allowlist, `DENYLIST` if item is on denylist

that said, even with this approach it seems like you'd still want to cross-reference the items with your own copy of the allowlist since you can't fully trust users haven't modified the allowlist you gave them (and if you only care about items on the allowlist, you don't need an expansive denylist)

**Comment by tmgrhm** [comment]
Really appreciate that, solution sounds solid to me, thank you so much.

Yes I will have a list further down the process which will verify the drop was valid (e.g. drop name matches a desired item name, all criteria for relevant tile being active are met, etc) but at the moment of receiving and processing the json, having the ability to filter out irrelevant items massively reduces clutter and processing overheads later in the process as it's quite expensive/slow to read cells in Google Apps Scripts.  
By having this enum I only need two copies of the item list (one in dink's allowlist for each participant, and another in the spreadsheet itself, one item per row, to keep track of each team's current progress) and can bypass having a third copy in the script to be referenced when receiving messages from dink.



Apologies if this is a GitHub faux pas, but while I have you, are you aware of any issues with `LOOT` seeing drops from Muspah? A friend had min value set to `1`, rarity to `0`, (i.e. should report every single drop) which is working fine for other mobs but it didn't trigger a single `LOOT` message from Muspah. I reproduced it too, but I know I've seen Venator Shards loot messages in discord previously. Loot tracker plugin did log the loot.  
I only tested once and can't be sure it's not an issue on my end, hence not opening a new issue yet, but I don't see why it wouldn't have triggered when it did so correctly for other mobs… I don't suppose you have any ideas? If not I'll do some further testing when I have time and open a new issue if need be.

Thanks again, hope you have a nice holiday :)
